perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java

```diff
@@ -839,17 +839,23 @@ public Object mapRow(ResultSet rs, int i) throws SQLException {
 
 	@Override
 	public List<Attribute> getAttributes(PerunSession sess, Facility facility) throws InternalErrorException {
+		List<Attribute> attributesCache = null;
+
 		if(!CacheManager.isCacheDisabled() && !perun.getCacheManager().wasCacheUpdatedInTransaction()) {
 			List<Attribute> attrs = perun.getCacheManager().getAllNonEmptyAttributes(new Holder(facility.getId(), Holder.HolderType.FACILITY));
-			return this.setValuesOfAttributes(sess, attrs, facility, null);
+			attributesCache = this.setValuesOfAttributes(sess, attrs, facility, null);
 		}
 
 		try {
-			return jdbc.query("select " + getAttributeMappingSelectQuery("fac") + " from attr_names " +
+			List<Attribute> attributesDB = jdbc.query("select " + getAttributeMappingSelectQuery("fac") + " from attr_names " +
 							"left join facility_attr_values fac    on id=fac.attr_id and fac.facility_id=? " +
 							"where namespace=? or (namespace in (?,?) and (attr_value is not null or attr_value_text is not null))",
 					new SingleBeanAttributeRowMapper<>(sess, this, facility), facility.getId(), AttributesManager.NS_FACILITY_ATTR_CORE, AttributesManager.NS_FACILITY_ATTR_DEF, AttributesManager.NS_FACILITY_ATTR_OPT);
 
+			String message = "The attributes do not match, attributes from cache are: " + attributesCache + " while attributes from db are: " + attributesDB + " the attributes belong to " + facility;
```

Comment:
This is even less effective than previously. Right now you serialize all objects all the time, not just when attributes doesn't match. You can pass string with placeholders and array of objects like This:

```java
// changed private method
private void compareAttributesFromCacheAndDB(List<Attribute> attributesCache, List<Attribute> attributesDB, String message, Object... messageParams)

// generic strings with placeholders
String message = "The attributes do not match, attributes from cache are: {} while attributes from db are: {} the attributes belong to {}";

// call of private method
compareAttributesFromCacheAndDB(attributesCache, attributesDB, message, facility);

// implementation in private method
log.error(message, attributesCache, attributesDb, messageParams);
```

You can probably even simplify it by having message fixed inside private methods and then care only about passed messageParams (which will be perun beans). It doesn't really matter if the text will end like: "_attributes belong to [[user], [facility]]_".
You could also append placeholders dynamically to the string base on the length of "messageParams" (as I mention in another comment). Then it wouldn't be placed as an array into the message. It might be necessary, if we want to pass exception as the last param to get stack trace from it.